src/cmap/mod.rs

```diff
@@ -57,38 +54,139 @@ impl From<Arc<ConnectionPoolInner>> for ConnectionPool {
+            address,
+            connection_options,
+        }
+    }
+
+    /// Emits an event from the event handler if one is present, where `emit` is a closure that uses
+    /// the event handler.
+    fn emit_event<F>(&self, emit: F)
+    where
+        F: FnOnce(&Arc<dyn CmapEventHandler>),
+    {
+        if let Some(handler) = self.connection_options.as_ref().and_then(|options| options.event_handler.as_ref()) {
+            emit(handler);
+        }
+    }
+
+    /// Fetches the next connection id, incrementing it for the next connection.
+    fn next_connection_id(&mut self) -> u32 {
+        let id = self.next_connection_id;
+        self.next_connection_id += 1;
+        id
+    }
+
+    /// Increments the generation of the pool. Rather than eagerly removing stale connections from
+    /// the pool, they are left for the background thread to clean up.
+    fn clear(&mut self) {
+        self.generation += 1;
+
+        self.emit_event(|handler| {
+            let event = PoolClearedEvent {
+                address: self.address.clone(),
+            };
+
+            handler.handle_pool_cleared_event(event);
+        });
+    }
+
+    /// Create a connection, incrementing the total connection count and emitting the appropriate
+    /// monitoring events.
+    fn create_connection(&mut self) -> Result<Connection> {
+        let mut connection = Connection::new(
+            self.next_connection_id(),
+            self.address.clone(),
+            self.generation,
+            self.connection_options.clone(),
+        )?;
+
+        self.emit_event(|handler| {
+            handler.handle_connection_created_event(connection.created_event())
+        });
+
+        let establish_result = self
+            .establisher
+            .establish_connection(&mut connection);
+
+        if let Err(e) = establish_result {
+            if e.is_authentication_error() {
+                // auth spec requires that the pool be cleared when encountering an auth error during establishment.
+                self.clear();
+            }
+            return Err(e);
+        }
+
+        self.total_connection_count += 1;
+        self.emit_event(|handler| handler.handle_connection_ready_event(connection.ready_event()));
+
+        Ok(connection)
+    }
+
+    /// Close a connection, emit the event for it being closed, and decrement the
+    /// total connection count.
+    fn close_connection(&mut self, connection: Connection, reason: ConnectionClosedReason) {
+        connection.close_and_drop(reason);
+        self.total_connection_count -= 1;
+    }
+}
+
+/// The internal state of a connection pool.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub(crate) struct ConnectionPoolInner {
+    /// The address the pool's connections will connect to.
+    address: StreamAddress,
+
+    /// The structure used to manage connection creation, closing, storage, and generation.
+    connection_manager: Arc<Mutex<ConnectionManager>>,
```

Comment:
this manager will only ever have a second reference to it when we move it to the task in `Drop` since we can't `await` on the mutex in a non-async function. It's unfortunate that we have two layers of `Arc` here, but it doesn't really seem like that big of a deal to me.

Reply:
We have two layers of `Arc` in other places in the driver, so I think it's fine.